lib/Frontend/ParseableInterfaceModuleLoader.cpp

```diff
@@ -0,0 +1,1022 @@
+                     diag::error_extracting_version_from_parseable_interface);
+      return true;
+    }
+    if (!FlagRe.match(SB, &FlagMatches)) {
+      diags.diagnose(diagnosticLoc,
+                     diag::error_extracting_flags_from_parseable_interface);
+      return true;
+    }
+    assert(VersMatches.size() == 2);
+    assert(FlagMatches.size() == 2);
+    Vers = swift::version::Version(VersMatches[1], SourceLoc(), &diags);
+    llvm::cl::TokenizeGNUCommandLine(FlagMatches[1], SubArgSaver, SubArgs);
+    return false;
+  }
+
+  /// Populate the provided \p Deps with \c FileDependency entries including:
+  ///
+  ///    - All the dependencies mentioned by \p SubInstance's DependencyTracker,
+  ///      that were read while compiling the module.
+  ///
+  ///    - For any file in the latter set that is itself a .swiftmodule
+  ///      living in \p cachePath, all of _its_ dependencies, copied
+  ///      out to avoid having to do recursive scanning when rechecking this
+  ///      dependency in the future.
+  bool collectDepsForSerialization(CompilerInstance &SubInstance,
+                                   SmallVectorImpl<FileDependency> &Deps,
+                                   bool IsHashBased) {
+    auto DTDeps = SubInstance.getDependencyTracker()->getDependencies();
+    SmallVector<StringRef, 16> InitialDepNames(DTDeps.begin(), DTDeps.end());
+    InitialDepNames.push_back(interfacePath);
+    llvm::StringSet<> AllDepNames;
+    for (auto const &DepName : InitialDepNames) {
+      if (AllDepNames.insert(DepName).second && dependencyTracker) {
+        dependencyTracker->addDependency(DepName, /*isSystem*/false);
+      }
+
+      auto DepBuf = getBufferOfDependency(fs, DepName, interfacePath,
+                                          diags, diagnosticLoc);
+      if (!DepBuf)
+        return true;
+      auto Status = getStatusOfDependency(fs, DepName, interfacePath,
+                                          diags, diagnosticLoc);
+      if (!Status)
+        return true;
+
+      if (IsHashBased) {
+        uint64_t hash = xxHash64(DepBuf->getBuffer());
+        Deps.push_back(
+          FileDependency::hashBased(DepName, Status->getSize(), hash));
+      } else {
+        uint64_t mtime =
+        Status->getLastModificationTime().time_since_epoch().count();
+        Deps.push_back(
+          FileDependency::modTimeBased(DepName, Status->getSize(), mtime));
+      }
+
+      if (moduleCachePath.empty())
+        continue;
+
+      // If Dep is itself a .swiftmodule in the cache dir, pull out its deps
+      // and include them in our own, so we have a single-file view of
+      // transitive deps: removes redundancies, and avoids opening and reading
+      // multiple swiftmodules during future loads.
+      auto Ext = llvm::sys::path::extension(DepName);
+      auto Ty = file_types::lookupTypeForExtension(Ext);
+      if (Ty == file_types::TY_SwiftModuleFile &&
+          DepName.startswith(moduleCachePath)) {
```

Comment:
That's an open question right now -- we don't want to inline the deps for prebuilt modules, otherwise the modules will ignore forwarding modules completely while validating dependencies, defeating the "only hash deps once" optimization.

rdar://48659199